### PR TITLE
Refine battle pathfinding using a recursive greedy depth-limited approach

### LINKED ISSUE #2951: Combat screen, 2-hex troops cannot reach moat hex and skip turn or start moving the wrong direction
When we place 2-hex troop (here Medusas) in a certain place inside castle:
Medusa problem
Medusa problem 2
And then we select a place to enter the moat - Medusas skip its turn and don't perform any action.

Edit:
The same happens here:
Medusa problem 5

Alos, if we place Medusas other place and try to reach the same cell - they move into wrong direction and stuck in a moat:
Medusa problem 3
Medusa problem 4

### PR DESCRIPTION
**Pull request by oleg-derevenetz**
Hi all,

I don't know if it's really worth it (considering the new (Dijkstra's based?) pathfinding engine from battle_pathfinding.cpp), but here is my proposal of greedy (in the sense of making the locally optimal choice at each stage) recursive pathfinding algorithm as a possible temporary replacement for old GetAStarPath(). This algorithm may be suboptimal in some situations, but it is simple and (at first sight) it shows much better results than old (in my opinion, weird or incorrectly implemented A* based) pathfinding algorithm, especially in the case of wide units.

Anyway, there is definitely one useful commit in this PR that fixes Board::GetDistance() function. Previous implementation gave incorrect results in a number of cases.

fix #2951

### CODE REVIEW

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , I will look at all pull requests very soon. I was absent for a few days.

**Comment by oleg-derevenetz** [comment]
Hi @ihhub Thank you, there is no hurry at all :) After all, weekends are made for relaxation :)

**Comment by ihhub** [comment]
Oh, no. @oleg-derevenetz , could you please fix merging confict here?

**Comment by oleg-derevenetz** [comment]
@ihhub Yes, of course. It's after merging of #2960.

**Comment by idshibanov** [comment]
I wonder if it's worth it to create a small suite of pathfinding test cases as we keep changing/fixing this code. It is challenging to test it (have to initialize Board and Arena instances), but could be useful to define it now.

> I don't know if it's really worth it (considering the new (Dijkstra's based?) pathfinding engine from battle_pathfinding.cpp)

New engine is indeed a Dijkstra based algorithm, focused on caching optimal moves to every cell rather than building a single path per request. It was written with AI in mind where it would evaluate every option available without extra calculations.

I could adjust it to make it work for player movement as well by setting cell direction and fixing buildPath call there; that's why I'm thinking it could be useful to compare all 3 implementations over same set of tasks.

**Comment by oleg-derevenetz** [comment]
Rebased this PR after merging of #2960 and resolved conflicts.

@idshibanov

> New engine is indeed a Dijkstra based algorithm, focused on caching optimal moves to every cell rather than building a single path per request. It was written with AI in mind where it would evaluate every option available without extra calculations.

Yeah, that's what I figured. Dijkstra based algorithm should work much more efficient, because it's complexity is about n^2, but it processes the entire field at a time, while complexity of recursive algorithm, although can be (almost) linear, can also be exponential  in certain edge cases.

Anyway, that was just an entertainment. The only 100% useful part of this is the fix for Board::GetDistance().

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , could we please then create a separate pull request just for `Board::GetDistance()` method fix while keeping this one open as I see that we have some concerns here?

**Comment by oleg-derevenetz** [comment]
Hi @ihhub I made a separate PR #3011 and I'll rebase this one once that PR will be merged to master branch.

**Comment by oleg-derevenetz** [comment]
Rebased & resolved merge conflicts.

**Comment by ihhub** [comment]
@Branikolog , could you please test these changes? We need to know how monsters are moving in the battle, especially corner cases such as Hydras.

**Comment by Branikolog** [comment]
@ihhub 
Okay!

**Comment by Branikolog** [comment]
Hi!
Found an issue:
Neutral creatures don't perform any action till we move any troop.
[Calm Zombies.zip](https://github.com/ihhub/fheroes2/files/6294770/Calm.Zombies.zip)
Attack Zombies, skip turn.

**Comment by oleg-derevenetz** [comment]
Hi all,

> Neutral creatures don't perform any action till we move any troop.

Interesting issue. The AI is sure that all Zombies are able to reach cell 1 (to attack Liches) and cell 90 (to attack Phoenixes) on the first turn (?), so it commands to move to cell 1 (Liches have a higher priority as a target), but there is no way to do it for Zombies because this cell is obviously too far, so Zombies doesn't move anywhere. However, AI doesn't use my code to calculate target availability, it uses code from battle_pathfinding.cpp (it calls Battle::Arena::CalculateMoveToUnit(), which uses it), which I didn't touch at all.

**Comment by Branikolog** [comment]
Ogres feel greatly encouraged, while standing, so they have their morale triggered! :)
image

**Comment by oleg-derevenetz** [comment]
@Branikolog There is clearly some inconsistency between planning of unit moves by AI and cells that are really reachable by those units, so when AI commands the unit to move to particular cell, there is no way for this unit to reach this cell (at least on the current turn). We need to "coordinate expectations and reality" somehow.

**Comment by Branikolog** [comment]
@oleg-derevenetz 
I see.
But in some cases troops can start moving... But not all...
image
Looks like only 1 stack of Pikemen was brave enough to rush enemy, while others are waiting for the result of this reckless attack...

**Comment by idshibanov** [comment]
Yes, main difference is that old pathfinder built a path to the cell that's outside of current unit's movement range and then cut it to appropriate length. AI is issuing a command to move where it is intending and Pathfinder made it valid.

We definitely have to handle it on either side. Old behavior was handy for AI programming since you don't have to validate every path, but I admit it is more prone to errors. I might make it use `BattlePathfinder::buildPath` method since it's very cheap to execute. This can be pushed in separate PR.

**Comment by oleg-derevenetz** [comment]
Hi @idshibanov

> I might make it use BattlePathfinder::buildPath method since it's very cheap to execute. This can be pushed in separate PR.

I think it would be the best solution.

**Comment by oleg-derevenetz** [comment]
Hi @Branikolog I just made the AI pathfinding and the new Battle::Board::GetPath() implementation interoperable, so AI-controlled units are no longer stuck if their target is unreachable on the current turn (and also rebased this PR to the latest version of the master branch). Could you please take a look to see if any AI-related issues remain (in terms of battle pathfinding)?

**Comment by Branikolog** [comment]
Hi, @oleg-derevenetz !
It's quite intriguing! 
I'll test your fix tomorrow, as I'm going to sleep now. :)

**Comment by oleg-derevenetz** [comment]
> I'll test your fix tomorrow, as I'm going to sleep now. :)

Thanks! Good night then! :)

**Comment by Branikolog** [comment]
Hi, @oleg-derevenetz !
I've started testing and found a small issue:
Player's creature move using not optimal path:
Druid walk

**Comment by oleg-derevenetz** [comment]
Hi @Branikolog

> Player's creature move using not optimal path:

Yes, this can happen. Path may be suboptimal - as long as it leads to the desired point. This is a feature of the algorithm, since it starts the search from the geometrically closest cells, and if the destination is reachable through those cells (even if the whole path is not the best one), then everything is OK.

**Comment by Branikolog** [comment]
@oleg-derevenetz 
I think, the game should choose the shortest path, as moving in a way we watch at screenshot looks quite weird.

**Comment by oleg-derevenetz** [comment]
@Branikolog As long as all cells that should be reachable for the given unit are in fact reachable, there is no issue. As I mentioned before, this algorithm may be suboptimal in certain cases (like this one), but at least it always finds a path to any reachable cell, even in complex cases (moat, wide units, etc). In due time, I decided to let it be suboptimal, but work faster. If someone have a better implementation, he is free to offer an appropriate PR :)

**Comment by oleg-derevenetz** [comment]
Hi @ihhub what do you think regarding this PR (in light of recent changes regarding AI interoperability)? BTW it also should fix the following error messages for AI-controlled units:

```
16:23:44: [DBG_BATTLE]	GetAStarPath:  Path is not found for Unit: [ 13 Unicorns, Blue, pos: 58, 57 ], destination: (head cell ID: 58, tail cell ID: 57)
16:23:44: [DBG_BATTLE]	ApplyActionAttack:  incorrect param: Unit: [ 22 Battle Dwarves, Blue, pos: 34, -1, mode(0x0), uid(0x00000ff3), speed(Average, 4), hp(423), die(3)) ] and Unit: [ 7 Royal Mummies, None, pos: 56, -1, reflect, mode(0x3), uid(0x00000ffc), speed(Standing, 0), hp(206), die(13)) ]
```

**Comment by ihhub** [comment]
> Hi @ihhub what do you think regarding this PR (in light of recent changes regarding AI interoperability)? BTW it also should fix error messages for AI-controlled units like these ones:
> 
> ```
> 16:23:44: [DBG_BATTLE]	GetAStarPath:  Path is not found for Unit: [ 13 Unicorns, Blue, pos: 58, 57 ], destination: (head cell ID: 58, tail cell ID: 57)
> 16:23:44: [DBG_BATTLE]	ApplyActionAttack:  incorrect param: Unit: [ 22 Battle Dwarves, Blue, pos: 34, -1, mode(0x0), uid(0x00000ff3), speed(Average, 4), hp(423), die(3)) ] and Unit: [ 7 Royal Mummies, None, pos: 56, -1, reflect, mode(0x3), uid(0x00000ffc), speed(Standing, 0), hp(206), die(13)) ]
> ```

My only concern is recursion. Ideally we should avoid it based on good programming style :) Let me please test a bit. I'm somehow busy these days with my own problems :(

**Comment by oleg-derevenetz** [comment]
> My only concern is recursion. Ideally we should avoid it based on good programming style :) Let me please test a bit. I'm somehow busy these days with my own problems :(

There is no hurry :) I agree that recursion may be dangerous, so I did my best to strongly cap it.

**Comment by ihhub** [comment]
> @Branikolog As long as all cells that should be reachable for the given unit are in fact reachable, there is no issue. As I mentioned before, this algorithm may be suboptimal in certain cases (like this one), but at least it always finds a path to any reachable cell, even in complex cases (moat, wide units, etc). In due time, I decided to let it be suboptimal, but work faster. If someone have a better implementation, he is free to offer an appropriate PR :)

What if we review the path found by this algorithm and check if we can move uint escaping extra cells. For the case mentioned by @Dranikolog: get the first cell, check all cells after second one and see if they are reachable. If yes, then remove all cells in between. Of course it reduces performance but a tiny performance drop should be okay here.

File: src/fheroes2/battle/battle_board.cpp [hunk 1]
```diff
@@ -204,261 +204,224 @@ void Battle::Board::SetScanPassability( const Unit & unit )
...
-                        cellMap[cellId].cost = cost + cellMap[currentCellId].cost;
-                    }
-                }
-            }
 
-            cellMap[currentCellId].open = false;
-            int32_t cost = MAXU16;
+    // Scan the available cells recursively in ascending order of distance
+    for ( const auto & cellCost : cellCosts ) {
+        int32_t cellId;
```
Comment (ihhub) [inline src/fheroes2/battle/battle_board.cpp:259]: I think we can easily do like: `const int32_t cellId = cellCost.second;`.

File: src/fheroes2/battle/battle_board.cpp [hunk 2]
```diff
@@ -204,261 +204,224 @@ void Battle::Board::SetScanPassability( const Unit & unit )
...
-            if ( result.size() > unit.GetSpeed() )
-                result.resize( unit.GetSpeed() );
+
+        // Calculate the distance from the cell in question to the destination, sort cells by distance
+        cellCosts.emplace( Board::GetDistance( headCellId, dstHeadCellId ) + Board::GetDistance( tailCellId, dstTailCellId ), headCellId );
+    }
+
+    // Scan the available cells recursively in ascending order of distance
+    for ( const auto & cellCost : cellCosts ) {
+        int32_t headCellId;
```
Comment (ihhub) [inline src/fheroes2/battle/battle_board.cpp:344]: I think we can easily do like: `const int32_t headCellId = cellCost.second;`.

**Review by ihhub — COMMENTED** [review]
Hi @oleg-derevenetz , I left few comments and also proposed a solution to cover @Branikolog 's concerns about non-optimal path. Would you mind to take a look?

**Comment by oleg-derevenetz** [comment]
Hi @ihhub this performance drop may be not so tiny, because suboptimal part of the path may not begin from the first cell. We will need to check all potential neighbors of all cells.

**Comment by ihhub** [comment]
> Hi @ihhub this performance drop may be not so tiny, because suboptimal part of the path may not begin from the first cell. We will need to check all potential neighbors of all cells.

But in comparison with A-start algorithm it should be still faster as A-star checks all cells.

**Comment by oleg-derevenetz** [comment]
> But in comparison with A-start algorithm it should be still faster as A-star checks all cells.

May be - in the best case. But in worst case multiple cells may be checked multiple times due to its recursive nature.

**Comment by oleg-derevenetz** [comment]
@ihhub BTW this optimization may just not work at all. Suppose that we have an obstacle in a form of 'C' letter. We can go around from the left side (longer) or from the right side (shorter), but these ways doesn't contain common cells, so if algorithm choose to use the left side, then we will never be able to "fix" it. If we always want to have a best path, then we should either do not make algorithm "greedy" (so it will turn into simple recursive, very suboptimal in terms of performance) or try to adapt other algorithm, like Dijkstra-based (like currently used by AI).

**Comment by ihhub** [comment]
> @ihhub BTW this optimization may just not work at all. Suppose that we have an obstacle in a form of 'C' letter. We can go around from the left side (longer) or from the right side (shorter), but these ways doesn't contain common cells, so if algorithm choose to use the left side, then we will never be able to "fix" it. If we always want to have a best path, then we should either do not make algorithm "greedy" (so it will turn into simple recursive, very suboptimal in terms of performance) or try to adapt other algorithm, like Dijkstra-based (like currently used by AI).

From user perspective we have to fix algorithm to use most optimal path otherwise many will ask why creatures move in such weird way. Right now the current implementation ha flaws. Even if the new algorithm will be slower it's still better as our aim is to have stable algorithm most of all.

**Comment by oleg-derevenetz** [comment]
@ihhub

> From user perspective we have to fix algorithm to use most optimal path otherwise many will ask why creatures move in such weird way

Unlike current issues with pathfinding, these weird ways doesn't break the gameplay at all. To get the best path using recursive approach (not "greedy" variant) you will need to check ALL variants of paths, it's ALWAYS n! complexity (IIRC). And it seems that Dijkstra doesn't currently work very well too.

**Comment by ihhub** [comment]
> @ihhub
> 
> > From user perspective we have to fix algorithm to use most optimal path otherwise many will ask why creatures move in such weird way
> 
> Unlike current issues with pathfinding, these weird ways doesn't break the gameplay at all. To get the best path using recursive approach (not "greedy" variant) you will need to check ALL variants of paths, it's ALWAYS n! complexity (IIRC). And it seems that Dijkstra doesn't currently work very well too.

Okay. Let's fix the most obvious problem with making the current chosen path more optimal if possible. I think this solution would be better than what we have right now. Doesn't matter if a monster goes around obstacle in different direction but at least it won't look like it walks extra cells. What do you think?

**Comment by oleg-derevenetz** [comment]
@ihhub OK, I'll try to add some "path squeezing" logic.

**Comment by oleg-derevenetz** [comment]
Hmm, it seems that it's not trivial in case of wide units (because they need to rotate themselves), I'll need to think a bit more...

File: src/fheroes2/ai/normal/ai_normal_battle.cpp [hunk 3]
```diff
@@ -678,6 +716,11 @@ namespace AI
 
     void Normal::BattleTurn( Arena & arena, const Unit & currentUnit, Actions & actions )
     {
+        Board * board = Arena::GetBoard();
```
[the remark this answers is not in the document]
Comment (oleg-derevenetz) [reply src/fheroes2/ai/normal/ai_normal_battle.cpp:719]: Let's do it in a separate PR then. I'd not like to clutter this one with unrelated changes - there are already a lot of tricky things to think about... :)

**Comment by oleg-derevenetz** [comment]
Hi @ihhub I implemented a sort of path straightening for regular units, please take a look when you'll have a free time. With wide units situation is a bit more complex, because they should traverse through the cells in a particular order to properly rotate themselves. But for the same reason they are less likely to detour.

File: src/fheroes2/ai/normal/ai_normal_battle.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ihhub) [reply src/fheroes2/ai/normal/ai_normal_battle.cpp:719]: Agreed.

**Comment by ihhub** [comment]
> Hi @ihhub I implemented a sort of path straightening for regular units, please take a look when you'll have a free time. With wide units situation is a bit more complex, because they should traverse through the cells in a particular order (and pass through the extra cells sometimes) to properly rotate themselves. But for the same reason they are less likely to detour.

Hm... I feel we're trying to reinvent some pathfinding algorithm by doing such fixes. How hard would it be to implement `Dijkstra` algorithm even without extra optimizations?

**Comment by oleg-derevenetz** [comment]
@ihhub

> Hm... I feel we're trying to reinvent some pathfinding algorithm by doing such fixes. How hard would it be to implement `Dijkstra` algorithm even without extra optimizations?

We are already have one - for AI. And it contains a ton of hacks related to wide units (as well as current A* implementation) and both of them still doesn't work properly. Wide units are the main issue here.

**Comment by oleg-derevenetz** [comment]
> @ihhub
> 
> > Hm... I feel we're trying to reinvent some pathfinding algorithm by doing such fixes. How hard would it be to implement `Dijkstra` algorithm even without extra optimizations?
> 
> We are already have one - for AI. And it contains a ton of hacks related to wide units (as well as current A* implementation, I even introduced some of these hacks myself) and both of them still doesn't work properly. Wide units are the main issue here.

**Comment by oleg-derevenetz** [comment]
The main problem with "classical" algorithms is that they are not adapted to the situation when an object occupies more than one cell of the graph, and all the cells occupied by it should move synchronously and in a certain order. Hacks inevitable, I suppose.

**Comment by idshibanov** [comment]
> We are already have one - for AI. And it contains a ton of hacks related to wide units (as well as current A* implementation, I even implemented some of these hacks myself, as I remember) and both of them still doesn't work properly. Wide units are the main issue here. Dealing with them is complicated.

I agree that hacks are inevitable - because wide units only move in certain directions, have to rotate and handle moat logic. I wonder what doesn't work in current AI pathfinder? As I can remember now it is missing `buildPath` implementation and doesn't check current unit speed allowing them to build paths outside of current range (needed for planning). Is there anything else?

**Comment by oleg-derevenetz** [comment]
Hi @idshibanov

> I agree that hacks are inevitable - because wide units only move in certain directions, have to rotate and handle moat logic. I wonder what doesn't work in current AI pathfinder? As I can remember now it is missing `buildPath` implementation and doesn't check current unit speed allowing them to build paths outside of current range (needed for planning). Is there anything else?

I tried it a few months ago and there were issues (IIRC, otherwise I'd use it for everything).

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , I think to solve our issues with normal and wide units we have to keep multiple copies of `Indexes` arrays on each level of recursion iteration. Then inside `for ( const auto & cellCost : cellCosts ) {` loop we populate each array and after return only one which has the smallest number of elements. This solution won't be optimal in terms on performance but I feel it'll solve the problem with non-optimal unit path.
Here is a draft for the idea (tested on wide units):
```
   std::vector<Indexes> tempResult( cellCosts.size(), result );
    std::vector<std::vector<bool>> tempVisitedCells( cellCosts.size(), visitedCells );

    bool someReachedDestination = false;
    std::vector<bool> reachedDestination( cellCosts.size(), false );

    // Scan the available cells recursively in ascending order of distance
    size_t index = 0;
    for ( const auto & cellCost : cellCosts ) {
        const int32_t headCellId = cellCost.second;

        // Mark the cell as visited for further steps
        tempVisitedCells[index].at( headCellId ) = true;

        // Turning back is not a movement
        const uint32_t steps = headCellId == currentTailCellId ? remainingSteps : remainingSteps - 1;

        if ( GetPathForWideUnit( unit, destination, steps, headCellId, currentHeadCellId, tempVisitedCells[index], tempResult[index] ) ) {
            tempResult[index].push_back( headCellId );
            reachedDestination[index] = true;
            someReachedDestination = true;

            ++index;
            continue;
        }

        // Unmark the cell as visited
        tempVisitedCells[index].at( headCellId ) = false;
        ++index;
    }

    if ( someReachedDestination ) {
        size_t shortest = 0;
        size_t shortestId = 0;

        for ( size_t i = 0; i < cellCosts.size(); ++i ) {
            if ( reachedDestination[i] ) {
                if ( shortest == 0 || shortest > tempResult[i].size() ) {
                    shortest = tempResult[i].size();
                    shortestId = i;
                }
            }
        }

        visitedCells = tempVisitedCells[shortestId];
        result = tempResult[shortestId];
        return true;
    }
    else if ( !cellCosts.empty() ) {
        size_t shortest = 0;
        size_t shortestId = 0;

        for ( size_t i = 0; i < cellCosts.size(); ++i ) {
            if ( shortest == 0 || shortest > tempResult[i].size() ) {
                shortest = tempResult[i].size();
                shortestId = i;
            }
        }

        visitedCells = tempVisitedCells[shortestId];
        result = tempResult[shortestId];
    }

    return false;
```

**Comment by oleg-derevenetz** [comment]
Hi @ihhub looks like another "check all paths and find the shortest one" algorithm with corresponding complexity issues. I suppose that performance issues may start pretty early - with fast units (like Champions with Haste) on slow computers delay might be noticeable. I'd not consider basic recursive algorithms here.

**Comment by ihhub** [comment]
> Hi @ihhub looks like another "check all paths and find the shortest one" algorithm with corresponding complexity issues. I suppose that performance issues may start pretty early - with fast units (like Champions with Haste) on slow computers delay might be noticeable. I'd not consider basic recursive algorithms here.

I absolutely agree with you about performance. That's why I wrote that my proposed solution is far from ideal. For example, we can optimize `tempResult` usage.

We can go with your implementation for now but later we still have to fix wide unit path.

**Comment by oleg-derevenetz** [comment]
@ihhub I don't think that noticeable issues with "weird paths" will be happen often, and, what's more important (IMHO) - they will not affect gameplay mechanics. Currently there are issues that units just are not able to reach a particular cell from another particular cell (IIRC, in OG there are similar issues also), which break mechanics.

**Comment by oleg-derevenetz** [comment]
@ihhub you can play with this algorithm yourself and evaluate it.

**Comment by ihhub** [comment]
@oleg-derevenetz , thank you so much for these changes! If someone will be unhappy we'll try to correct the algorithm later :)
